File: cpp/tests/unit-tests/MutableMessageFactoryTest.cpp
/*
 * #%L
 * %%
 * Copyright (C) 2017 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
#include <chrono>
#include <cstdint>
#include <string>

#include <gtest/gtest.h>

#include "joynr/Message.h"
#include "joynr/MessagingQos.h"
#include "joynr/MulticastPublication.h"
#include "joynr/MulticastSubscriptionQos.h"
#include "joynr/MulticastSubscriptionRequest.h"
#include "joynr/MutableMessage.h"
#include "joynr/MutableMessageFactory.h"
#include "joynr/OnChangeSubscriptionQos.h"
#include "joynr/Reply.h"
#include "joynr/Request.h"
#include "joynr/SubscriptionPublication.h"
#include "joynr/SubscriptionReply.h"
#include "joynr/SubscriptionRequest.h"
#include "joynr/SubscriptionStop.h"
#include "joynr/TimePoint.h"

using namespace joynr;

class MutableMessageFactoryTest : public ::testing::Test
{
public:
    MutableMessageFactoryTest()
            : messageFactory(0),
              senderID(),
              receiverID(),
              requestReplyID(),
              isLocalMessage(true),
              qos(),
              request(),
              oneWayRequest(),
              reply(),
              subscriptionPublication()
    {
    }

    void SetUp()
    {
        senderID = std::string("senderId");
        receiverID = std::string("receiverID");
        requestReplyID = "requestReplyID";
        qos = MessagingQos(456000);
        qos.putCustomMessageHeader("test-header", "test-header-value");
        request.setMethodName("methodName");
        request.setRequestReplyId(requestReplyID);
        request.setParams(42, "value");
        request.setParamDatatypes({"java.lang.Integer", "java.lang.String"});
        oneWayRequest.setMethodName("methodName");
        oneWayRequest.setParams(42, "value");
        oneWayRequest.setParamDatatypes({"java.lang.Integer", "java.lang.String"});
        reply.setRequestReplyId(requestReplyID);
        reply.setResponse("response");

        std::string subscriptionId("subscriptionTestId");
        subscriptionPublication.setSubscriptionId(subscriptionId);
        subscriptionPublication.setResponse("publication");
    }

    void checkSenderRecipient(const MutableMessage& mutableMessage)
    {
        EXPECT_EQ(senderID, mutableMessage.getSender());
        EXPECT_EQ(receiverID, mutableMessage.getRecipient());
        boost::optional<std::string> customHeader = mutableMessage.getCustomHeader("test-header");
        ASSERT_TRUE(customHeader.is_initialized());
        EXPECT_EQ("test-header-value", *customHeader);
    }

    void checkRequest(const MutableMessage& mutableMessage)
    {
        std::stringstream expectedPayloadStream;
        expectedPayloadStream << R"({"_typeName":"joynr.Request",)";
        expectedPayloadStream << R"("methodName":"methodName",)";
        expectedPayloadStream << R"("paramDatatypes":["java.lang.Integer","java.lang.String"],)";
        expectedPayloadStream << R"("params":[42,"value"],)";
        expectedPayloadStream << R"("requestReplyId":")" << request.getRequestReplyId() << R"("})";
        std::string expectedPayload = expectedPayloadStream.str();
        EXPECT_EQ(expectedPayload, mutableMessage.getPayload());
    }

    void checkOneWayRequest(const MutableMessage& mutableMessage)
    {
        std::stringstream expectedPayloadStream;
        expectedPayloadStream << R"({"_typeName":"joynr.OneWayRequest",)";
        expectedPayloadStream << R"("methodName":"methodName",)";
        expectedPayloadStream << R"("paramDatatypes":["java.lang.Integer","java.lang.String"],)";
        expectedPayloadStream << R"("params":[42,"value"])";
        expectedPayloadStream << R"(})";
        std::string expectedPayload = expectedPayloadStream.str();
        EXPECT_EQ(expectedPayload, mutableMessage.getPayload());
    }

    void checkReply(const MutableMessage& mutableMessage)
    {
        std::stringstream expectedPayloadStream;
        expectedPayloadStream << R"({"_typeName":"joynr.Reply",)";
        expectedPayloadStream << R"("response":["response"],)";
        expectedPayloadStream << R"("requestReplyId":")" << reply.getRequestReplyId() << R"("})";
        std::string expectedPayload = expectedPayloadStream.str();
        EXPECT_EQ(expectedPayload, mutableMessage.getPayload());
    }

    void checkSubscriptionPublication(const MutableMessage& mutableMessage)
    {
        std::stringstream expectedPayloadStream;
        expectedPayloadStream << R"({"_typeName":"joynr.SubscriptionPublication",)";
        expectedPayloadStream << R"("response":["publication"],)";
        expectedPayloadStream << R"("subscriptionId":")"
                              << subscriptionPublication.getSubscriptionId() << R"("})";
        std::string expectedPayload = expectedPayloadStream.str();
        EXPECT_EQ(expectedPayload, mutableMessage.getPayload());
    }

    void checkMulticastPublication(const MutableMessage& mutableMessage,
                                   const MulticastPublication& multicastPublication)
    {
        std::stringstream expectedPayloadStream;
        expectedPayloadStream << R"({"_typeName":"joynr.MulticastPublication",)";
        expectedPayloadStream << R"("response":["publication"],)";
        expectedPayloadStream << R"("multicastId":")" << multicastPublication.getMulticastId()
                              << R"("})";
        std::string expectedPayload = expectedPayloadStream.str();
        EXPECT_EQ(expectedPayload, mutableMessage.getPayload());
    }

protected:
    ADD_LOGGER(MutableMessageFactoryTest)
    MutableMessageFactory messageFactory;
    std::string senderID;
    std::string receiverID;
    std::string requestReplyID;
    const bool isLocalMessage;
    MessagingQos qos;
    Request request;
    OneWayRequest oneWayRequest;
    Reply reply;
    SubscriptionPublication subscriptionPublication;
};

TEST_F(MutableMessageFactoryTest, createRequest_withContentType)
{
    MutableMessage mutableMessage =
            messageFactory.createRequest(senderID, receiverID, qos, request, isLocalMessage);
    checkSenderRecipient(mutableMessage);
}

TEST_F(MutableMessageFactoryTest, createRequest)
{
    MutableMessage mutableMessage =
            messageFactory.createRequest(senderID, receiverID, qos, request, isLocalMessage);
    const TimePoint expectedExpiryDate = TimePoint::fromRelativeMs(static_cast<std::int64_t>(qos.getTtl()));
    const TimePoint expiryDate = mutableMessage.getExpiryDate();
    EXPECT_NEAR(expectedExpiryDate.toMilliseconds(), expiryDate.toMilliseconds(), 100.);
    JOYNR_LOG_DEBUG(
            logger(), "expiryDate: {} [{}]", expiryDate.toString(), expiryDate.toMilliseconds());
    JOYNR_LOG_DEBUG(logger(),
                    "expectedExpiryDate: {}  [{}]",
                    expectedExpiryDate.toString(),
                    expectedExpiryDate.toMilliseconds());

    checkSenderRecipient(mutableMessage);
    checkRequest(mutableMessage);
    EXPECT_EQ(Message::VALUE_MESSAGE_TYPE_REQUEST(), mutableMessage.getType());
}

TEST_F(MutableMessageFactoryTest, createReply)
{
    MutableMessage mutableMessage =
            messageFactory.createReply(senderID, receiverID, qos, {}, reply);
    checkSenderRecipient(mutableMessage);
    checkReply(mutableMessage);
    EXPECT_EQ(Message::VALUE_MESSAGE_TYPE_REPLY(), mutableMessage.getType());
}

TEST_F(MutableMessageFactoryTest, createOneWayRequest)
{
    MutableMessage mutableMessage = messageFactory.createOneWayRequest(
            senderID, receiverID, qos, oneWayRequest, isLocalMessage);
    checkSenderRecipient(mutableMessage);
    checkOneWayRequest(mutableMessage);
    EXPECT_EQ(Message::VALUE_MESSAGE_TYPE_ONE_WAY(), mutableMessage.getType());
}

TEST_F(MutableMessageFactoryTest, createPublication)
{
    MutableMessage mutableMessage = messageFactory.createSubscriptionPublication(
            senderID, receiverID, qos, subscriptionPublication);
    checkSenderRecipient(mutableMessage);
    checkSubscriptionPublication(mutableMessage);
    EXPECT_EQ(Message::VALUE_MESSAGE_TYPE_PUBLICATION(), mutableMessage.getType());
}

TEST_F(MutableMessageFactoryTest, createMulticastPublication)
{
    MulticastPublication multicastPublication;
    multicastPublication.setMulticastId(receiverID);
    multicastPublication.setResponse("publication");
    MutableMessage mutableMessage =
            messageFactory.createMulticastPublication(senderID, qos, multicastPublication);
    checkSenderRecipient(mutableMessage);
    checkMulticastPublication(mutableMessage, multicastPublication);
    EXPECT_EQ(Message::VALUE_MESSAGE_TYPE_MULTICAST(), mutableMessage.getType());
}

TEST_F(MutableMessageFactoryTest, createSubscriptionRequest)
{
    auto subscriptionQos = std::make_shared<OnChangeSubscriptionQos>();
    SubscriptionRequest subscriptionRequest;
    subscriptionRequest.setSubscriptionId("subscriptionId");
    subscriptionRequest.setSubscribeToName("attributeName");
    subscriptionRequest.setQos(subscriptionQos);
    MutableMessage mutableMessage = messageFactory.createSubscriptionRequest(
            senderID, receiverID, qos, subscriptionRequest, isLocalMessage);
    checkSenderRecipient(mutableMessage);
    EXPECT_EQ(Message::VALUE_MESSAGE_TYPE_SUBSCRIPTION_REQUEST(), mutableMessage.getType());
}

TEST_F(MutableMessageFactoryTest, createMulticastSubscriptionRequest)
{
    auto subscriptionQos = std::make_shared<MulticastSubscriptionQos>();
    MulticastSubscriptionRequest subscriptionRequest;
    subscriptionRequest.setSubscriptionId("subscriptionId");
    subscriptionRequest.setMulticastId("multicastId");
    subscriptionRequest.setSubscribeToName("attributeName");
    subscriptionRequest.setQos(subscriptionQos);
    MutableMessage mutableMessage = messageFactory.createMulticastSubscriptionRequest(
            senderID, receiverID, qos, subscriptionRequest, isLocalMessage);
    checkSenderRecipient(mutableMessage);
    EXPECT_EQ(
            Message::VALUE_MESSAGE_TYPE_MULTICAST_SUBSCRIPTION_REQUEST(), mutableMessage.getType());
}

TEST_F(MutableMessageFactoryTest, createSubscriptionStop)
{
    std::string subscriptionId("TEST-SubscriptionId");
    SubscriptionStop subscriptionStop;
    subscriptionStop.setSubscriptionId(subscriptionId);
    MutableMessage mutableMessage =
            messageFactory.createSubscriptionStop(senderID, receiverID, qos, subscriptionStop);
    checkSenderRecipient(mutableMessage);
    EXPECT_EQ(Message::VALUE_MESSAGE_TYPE_SUBSCRIPTION_STOP(), mutableMessage.getType());
}

TEST_F(MutableMessageFactoryTest, testSetNoEffortHeader)
{
    MutableMessage mutableMessage =
            messageFactory.createRequest(senderID, receiverID, qos, request, isLocalMessage);
    const boost::optional<std::string> optionalEffort = mutableMessage.getEffort();
    EXPECT_FALSE(optionalEffort);
}

TEST_F(MutableMessageFactoryTest, testSetBestEffortHeader)
{
    qos.setEffort(MessagingQosEffort::Enum::BEST_EFFORT);
    MutableMessage mutableMessage =
            messageFactory.createRequest(senderID, receiverID, qos, request, isLocalMessage);
    const boost::optional<std::string> optionalEffort = mutableMessage.getEffort();
    ASSERT_TRUE(optionalEffort.is_initialized());
    EXPECT_EQ(
            MessagingQosEffort::getLiteral(MessagingQosEffort::Enum::BEST_EFFORT), *optionalEffort);
}

TEST_F(MutableMessageFactoryTest, compressFlagIsPropagated)
{
    const bool compress = true;
    qos.setCompress(compress);
    MutableMessage mutableMessage =
            messageFactory.createRequest(senderID, receiverID, qos, request, isLocalMessage);
    EXPECT_EQ(compress, mutableMessage.getCompress());
}

File: java/core/libjoynr/src/test/java/io/joynr/dispatching/MutableMessageFactoryTest.java
/*
 * #%L
 * %%
 * Copyright (C) 2020 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package io.joynr.dispatching;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

import java.lang.reflect.Method;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.Properties;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.junit.MockitoJUnitRunner;

import com.fasterxml.jackson.databind.JsonNode;
import com.google.inject.AbstractModule;
import com.google.inject.Guice;
import com.google.inject.Injector;
import com.google.inject.TypeLiteral;
import com.google.inject.multibindings.Multibinder;
import com.google.inject.name.Names;

import io.joynr.common.ExpiryDate;
import io.joynr.common.JoynrPropertiesModule;
import io.joynr.messaging.ConfigurableMessagingSettings;
import io.joynr.messaging.JoynrMessageProcessor;
import io.joynr.messaging.JsonMessageSerializerModule;
import io.joynr.messaging.MessagingQos;
import io.joynr.messaging.MessagingQosEffort;
import io.joynr.util.ObjectMapper;
import joynr.ImmutableMessage;
import joynr.Message;
import joynr.MulticastPublication;
import joynr.MutableMessage;
import joynr.PeriodicSubscriptionQos;
import joynr.Reply;
import joynr.Request;
import joynr.SubscriptionPublication;
import joynr.SubscriptionRequest;

@RunWith(MockitoJUnitRunner.class)
public class MutableMessageFactoryTest {
    private static final long TTL = 1000;
    private static final long MAX_ALLOWED_EXPIRY_DATE_DIFF_MS = 500;
    private static final long NO_TTL_UPLIFT = 0;
    private MutableMessageFactory mutableMessageFactory;
    private String fromParticipantId;
    private String toParticipantId;
    private Request request;
    private Reply reply;
    private String payload;
    private ExpiryDate expiryDate;
    private MessagingQos messagingQos;
    private SubscriptionRequest subscriptionRequest;

    private SubscriptionPublication publication;
    private ObjectMapper objectMapper;

    @Before
    public void setUp() throws NoSuchMethodException, SecurityException {
        fromParticipantId = "sender";
        toParticipantId = "receiver";

        Injector injector = Guice.createInjector(new JoynrPropertiesModule(new Properties()),
                                                 new JsonMessageSerializerModule(),
                                                 new AbstractModule() {

                                                     @Override
                                                     protected void configure() {
                                                         bind(Long.class).annotatedWith(Names.named(ConfigurableMessagingSettings.PROPERTY_TTL_UPLIFT_MS))
                                                                         .toInstance(NO_TTL_UPLIFT);
                                                         requestStaticInjection(Request.class);
                                                         Multibinder<JoynrMessageProcessor> joynrMessageProcessorMultibinder = Multibinder.newSetBinder(binder(),
                                                                                                                                                        new TypeLiteral<JoynrMessageProcessor>() {
                                                                                                                                                        });
                                                         joynrMessageProcessorMultibinder.addBinding()
                                                                                         .toInstance(new JoynrMessageProcessor() {
                                                                                             @Override
                                                                                             public MutableMessage processOutgoing(MutableMessage joynrMessage) {
                                                                                                 joynrMessage.getCustomHeaders()
                                                                                                             .put("test",
                                                                                                                  "test");
                                                                                                 return joynrMessage;
                                                                                             }

                                                                                             @Override
                                                                                             public ImmutableMessage processIncoming(ImmutableMessage joynrMessage) {
                                                                                                 return joynrMessage;
                                                                                             }
                                                                                         });
                                                     }
                                                 });

        objectMapper = injector.getInstance(ObjectMapper.class);

        payload = "payload";
        Method method = TestProvider.class.getMethod("methodWithStrings", new Class[]{ String.class });
        request = new Request(method.getName(), new String[]{ payload }, method.getParameterTypes());
        String requestReplyId = request.getRequestReplyId();
        reply = new Reply(requestReplyId, objectMapper.<JsonNode> valueToTree(payload));
        messagingQos = new MessagingQos(TTL);
        expiryDate = ExpiryDate.fromRelativeTtl(messagingQos.getRoundTripTtl_ms());

        String subscriptionId = "subscription";
        String attributeName = "attribute";
        PeriodicSubscriptionQos subscriptionqos = new PeriodicSubscriptionQos();
        subscriptionqos.setPeriodMs(1000).setValidityMs(10).setAlertAfterIntervalMs(1500).setPublicationTtlMs(1000);
        subscriptionRequest = new SubscriptionRequest(subscriptionId, attributeName, subscriptionqos);
        String response = "response";
        publication = new SubscriptionPublication(Arrays.asList(response), subscriptionId);

        mutableMessageFactory = injector.getInstance(MutableMessageFactory.class);
    }

    public static void assertExpiryDateEquals(long expectedValue, MutableMessage message) {
        long diff = Math.abs(expectedValue - message.getTtlMs());

        assertTrue(message.isTtlAbsolute());
        assertTrue("ExpiryDate=" + message.getTtlMs() + " differs " + diff + "ms (more than "
                + MAX_ALLOWED_EXPIRY_DATE_DIFF_MS + "ms) from expected value=" + expectedValue,
                   diff <= MAX_ALLOWED_EXPIRY_DATE_DIFF_MS);
    }

    @Test
    public void createRequest() {
        MutableMessage message = mutableMessageFactory.createRequest(fromParticipantId,
                                                                     toParticipantId,
                                                                     request,
                                                                     messagingQos);
        assertEquals(Message.MessageType.VALUE_MESSAGE_TYPE_REQUEST, message.getType());

        assertEquals(fromParticipantId, message.getSender());
        assertEquals(toParticipantId, message.getRecipient());

        assertExpiryDateEquals(expiryDate.getValue(), message);
        assertTrue(message.getPayload() != null);
    }

    @Test
    public void createRequestWithCustomEffort() {
        MessagingQos customMessagingQos = new MessagingQos();
        customMessagingQos.setEffort(MessagingQosEffort.BEST_EFFORT);
        MutableMessage message = mutableMessageFactory.createRequest(fromParticipantId,
                                                                     toParticipantId,
                                                                     request,
                                                                     customMessagingQos);
        expiryDate = ExpiryDate.fromRelativeTtl(customMessagingQos.getRoundTripTtl_ms());
        assertExpiryDateEquals(expiryDate.getValue(), message);
        assertEquals(String.valueOf(MessagingQosEffort.BEST_EFFORT), message.getEffort());
    }

    @Test
    public void createRequestWithCustomHeaders() throws Exception {
        final Map<String, String> myCustomHeaders = new HashMap<>();
        final String headerName = "header";
        final String headerValue = "value";
        myCustomHeaders.put(headerName, headerValue);
        messagingQos.getCustomMessageHeaders().putAll(myCustomHeaders);
        MutableMessage message = mutableMessageFactory.createRequest(fromParticipantId,
                                                                     toParticipantId,
                                                                     request,
                                                                     messagingQos);
        assertEquals(Message.MessageType.VALUE_MESSAGE_TYPE_REQUEST, message.getType());
        assertExpiryDateEquals(expiryDate.getValue(), message);

        final String expectedCustomHeaderName = Message.CUSTOM_HEADER_PREFIX + headerName;
        assertTrue(message.getImmutableMessage().getHeaders().containsKey(expectedCustomHeaderName));

        Map<String, String> customHeaders = message.getCustomHeaders();
        assertTrue(customHeaders.size() == 3);
        assertTrue(customHeaders.containsKey(headerName));
    }

    @Test
    public void testCreateOneWayRequest() {
        MutableMessage message = mutableMessageFactory.createOneWayRequest(fromParticipantId,
                                                                           toParticipantId,
                                                                           request,
                                                                           messagingQos);
        assertNotNull(message);
        assertEquals(Message.MessageType.VALUE_MESSAGE_TYPE_ONE_WAY, message.getType());
        assertEquals(fromParticipantId, message.getSender());
        assertEquals(toParticipantId, message.getRecipient());
        assertExpiryDateEquals(expiryDate.getValue(), message);
        assertNotNull(message.getPayload());
    }

    @Test
    public void testCreateOneWayRequestWithCustomHeader() throws Exception {
        final Map<String, String> myCustomHeaders = new HashMap<>();
        final String headerName = "header";
        final String headerValue = "value";
        myCustomHeaders.put(headerName, headerValue);
        messagingQos.getCustomMessageHeaders().putAll(myCustomHeaders);
        MutableMessage message = mutableMessageFactory.createOneWayRequest(fromParticipantId,
                                                                           toParticipantId,
                                                                           request,
                                                                           messagingQos);
        assertNotNull(message);
        assertEquals(Message.MessageType.VALUE_MESSAGE_TYPE_ONE_WAY, message.getType());
        assertExpiryDateEquals(expiryDate.getValue(), message);

        final String expectedCustomHeaderName = Message.CUSTOM_HEADER_PREFIX + headerName;
        assertTrue(message.getImmutableMessage().getHeaders().containsKey(expectedCustomHeaderName));

        Map<String, String> customHeaders = message.getCustomHeaders();
        assertTrue(customHeaders.size() == 2);
        assertTrue(customHeaders.containsKey(headerName));
    }

    @Test
    public void createReply() {
        MutableMessage message = mutableMessageFactory.createReply(fromParticipantId,
                                                                   toParticipantId,
                                                                   reply,
                                                                   messagingQos);

        assertEquals(Message.MessageType.VALUE_MESSAGE_TYPE_REPLY, message.getType());
        assertEquals(fromParticipantId, message.getSender());
        assertEquals(toParticipantId, message.getRecipient());
        assertExpiryDateEquals(expiryDate.getValue(), message);

        assertTrue(message.getPayload() != null);
    }

    @Test
    public void createSubscriptionRequest() {
        MutableMessage message = mutableMessageFactory.createSubscriptionRequest(fromParticipantId,
                                                                                 toParticipantId,
                                                                                 subscriptionRequest,
                                                                                 messagingQos);
        assertEquals(Message.MessageType.VALUE_MESSAGE_TYPE_SUBSCRIPTION_REQUEST, message.getType());
        assertEquals(fromParticipantId, message.getSender());
        assertEquals(toParticipantId, message.getRecipient());
        assertExpiryDateEquals(expiryDate.getValue(), message);

        assertTrue(message.getPayload() != null);
    }

    @Test
    public void createPublication() {
        MutableMessage message = mutableMessageFactory.createPublication(fromParticipantId,
                                                                         toParticipantId,
                                                                         publication,
                                                                         messagingQos);
        assertEquals(Message.MessageType.VALUE_MESSAGE_TYPE_PUBLICATION, message.getType());
        assertEquals(fromParticipantId, message.getSender());
        assertEquals(toParticipantId, message.getRecipient());
        assertExpiryDateEquals(expiryDate.getValue(), message);

        assertTrue(message.getPayload() != null);
    }

    @Test
    public void testMessageProcessorUsed() {
        MutableMessage message = mutableMessageFactory.createRequest("from",
                                                                     "to",
                                                                     new Request("name", new Object[0], new Class[0]),
                                                                     new MessagingQos());
        assertNotNull(message.getCustomHeaders().get("test"));
        assertEquals("test", message.getCustomHeaders().get("test"));
    }

    @Test
    public void testCreateMulticastMessage() {
        String multicastId = "multicastId";
        MulticastPublication multicastPublication = new MulticastPublication(Collections.emptyList(), multicastId);

        MutableMessage joynrMessage = mutableMessageFactory.createMulticast(fromParticipantId,
                                                                            multicastPublication,
                                                                            messagingQos);

        assertNotNull(joynrMessage);
        assertExpiryDateEquals(expiryDate.getValue(), joynrMessage);
        assertEquals(fromParticipantId, joynrMessage.getSender());
        assertEquals(multicastId, joynrMessage.getRecipient());
        assertEquals(Message.MessageType.VALUE_MESSAGE_TYPE_MULTICAST, joynrMessage.getType());
        assertTrue(new String(joynrMessage.getPayload(),
                              StandardCharsets.UTF_8).contains(MulticastPublication.class.getName()));
    }

    @Test
    public void testCompressedFlagIsSet() {
        MutableMessage message;
        MessagingQos messagingQos = new MessagingQos();

        messagingQos.setCompress(true);
        message = mutableMessageFactory.createRequest("from", "to", request, messagingQos);
        assertEquals(true, message.getCompressed());

        messagingQos.setCompress(false);
        message = mutableMessageFactory.createRequest("from", "to", request, messagingQos);
        assertEquals(false, message.getCompressed());
    }
}
